File: src/rda_python_setuid/pywrapper.c
/***************************************************************************************\
 *
 *    Title: pywrapper.c
 *   Author: Zaihua Ji, zji@ucar.edu
 *     Date: 2025-03-17
 *  Purpose: C wrapper to start any executable python code as an effective user
 *
 * Instruction:
 *    python -m venv $ENVHOME
 *    python -m pip install rda_python_setuid
 *
 *    # Compile pywrapper and set setuid bit (owned by CommonUser):
 *    pywrapper-install --user CommonUser [--envhome $ENVHOME/bin]
 *
 *    # For an existing python program, $ENVHOME/bin/CommonProgram.py, to execute it
 *    # as the common user (Mode 1 - symlink):
 *    pywrapper-install --link CommonProgram [--envhome $ENVHOME/bin]
 *    CommonProgram [options]
 *
 *    # For a specialist to run commands as themselves via pgstart (Mode 2):
 *    pywrapper-install --pgstart --user EffectUser [--envhome $ENVHOME/bin]
 *    pgstart_EffectUser EffectProgram [options]
 *
 *           N: python 3 release number, it is 10 for Python 3.10.12
 *  CommonUser: a common user login name, such as gdexdata, for GDEXMS configuration
 *  EffectUser: any user login name in the same group of the common user
 *    $ENVHOME: /glade/u/home/gdexdata/rdamsenv (venv) on DECS machines, and
 *              /glade/work/gdexdata/conda-envs/pg-rda (conda) on DAV
 *
 * Convention:
 *    Any Python package whose program is to be run via pywrapper must register
 *    its connector entry point with a setuid_ prefix in pyproject.toml, e.g.:
 *       [project.scripts]
 *       "setuid_dsarch" = "rda_python_dsarch.dsarch:main"
 *    pip install places setuid_dsarch in $ENVHOME/bin/.  pywrapper locates it by
 *    prepending setuid_ to the invoked program name.  No manual script creation needed.
 *    pywrapper-install --link will chown setuid_dsarch to CommonUser and chmod 700,
 *    preventing direct execution by other users while still allowing pywrapper
 *    (which runs with EUID=CommonUser via the setuid bit) to execv it.
 *
 \***************************************************************************************/

#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <limits.h>
#include <libgen.h>

int is_executable(const char *filename) {
   struct stat sb;

   if(stat(filename, &sb) == 0 && sb.st_mode & S_IXUSR) {
      return 1;
   } else {
      return 0;
   }
}

/* main program */
int main(int argc, char *argv[]) {
   char *name;
   char cname[80], prog[PATH_MAX];
   char exepath[PATH_MAX];
   char *fpath;
   ssize_t exelen;
   char pgstart[] = "pgstart";
   char **apntr = argv;

   exelen = readlink("/proc/self/exe", exepath, sizeof(exepath) - 1);
   if(exelen == -1) {
      perror("readlink /proc/self/exe");
      exit(1);
   }
   exepath[exelen] = '\0';
   fpath = dirname(exepath);

   name = strrchr(argv[0], '/');
   strncpy(cname, (name == NULL ? argv[0] : ++name), sizeof(cname) - 1);
   cname[sizeof(cname) - 1] = '\0';

   if(strstr(cname, pgstart) == cname) {
      if(argc == 1 || argv[1][0] == '-') {
         strncpy(cname, pgstart, sizeof(cname) - 1);
      } else {
         argv += 1;
         name = strrchr(argv[0], '/');
         strncpy(cname, (name == NULL ? argv[0] : ++name), sizeof(cname) - 1);
         cname[sizeof(cname) - 1] = '\0';
      }
   }
   if(snprintf(prog, sizeof(prog), "%s/setuid_%s", fpath, cname) >= (int)sizeof(prog)) {
      fprintf(stderr, "pywrapper: path too long\n");
      exit(1);
   }
   if(is_executable(prog) == 0) {
      /* fall back to cname.py for backward compatibility */
      if(snprintf(prog, sizeof(prog), "%s/%s.py", fpath, cname) >= (int)sizeof(prog)) {
         fprintf(stderr, "pywrapper: path too long\n");
         exit(1);
      }
      if(is_executable(prog) == 0) {
         if(snprintf(prog, sizeof(prog), "%s/pgstart.py", fpath) >= (int)sizeof(prog)) {
            fprintf(stderr, "pywrapper: path too long\n");
            exit(1);
         }
         argv = apntr;
      }
   }
   execv(prog, argv);  /* call Python script */
   perror(prog);       /* execv only returns on error */
   exit(1);
}

File: src/rda_python_setuid/pywrapper.py
#!/usr/bin/env python3
#
##################################################################################
#
#     Title: pywrapper
#    Author: Zaihua Ji, zji@ucar.edu
#      Date: 10/17/2020
#            2025-03-18 transferred to package rda_python_setuid from
#            https://github.com/NCAR/rda-utility-programs.git
#   Purpose: default python script for pywrapper.c program to show the instruction
#            of how to wrapper a python script under pywrapper
#
#    Github: https://github.com/NCAR/rda-python-setuid.git
#
##################################################################################

import os
import sys
import pwd
from rda_python_common.pg_log import PgLOG

def main():
   """Entry point for setuid_pywrapper (the default pywrapper.c fallback target).

   When pywrapper.c cannot find a matching setuid_<program> entry point, it
   falls back to executing this script.  Prints UID information and setup
   instructions, or dumps diagnostic information when a flag is supplied.

   Options:
      -env  -- print environment variables and exit
      -inc  -- print sys.path include paths and exit
      -plg  -- print PGLOG variables and exit
   """
   pglog = PgLOG()
   pglog.set_suid(pglog.PGLOG['EUID'])
   inc = True
   print("********************************************************************")
   argv = sys.argv[1:]
   if argv:
      if argv[0] == "-env":
         print("Environment Variables:")
         for ename in sorted(os.environ):
            print("{}: {}".format(ename, os.environ[ename]))
         inc = False
      elif argv[0] == "-inc":
         print("Including Paths:")
         for pname in sorted(sys.path):
            print(pname)
         inc = False
      elif argv[0] == "-plg":
         print("PGLOG variables:")
         for vname in sorted(pglog.PGLOG):
            print("{}: {}".format(vname, pglog.PGLOG[vname]))
         inc = False
      else:
         print("* {}: Unknown option".format(argv[0]))
   else:
      ruid = pglog.PGLOG['RUID']
      euid = pglog.PGLOG['EUID']
      ruser = pwd.getpwuid(ruid).pw_name
      euser = pwd.getpwuid(euid).pw_name
      print("* Your Login Name is {}({}) & Effective User Name is {}({}).".format(ruser, ruid, euser, euid))
      print("* To wrap a Python program 'yourscript', register its entry point")
      print("* with a setuid_ prefix in pyproject.toml:")
      print("*    [project.scripts]")
      print("*    \"setuid_yourscript\" = \"your_package.module:main\"")
      print("* Then run: pywrapper-install -l yourscript")
   if inc:
      print("* Include -env to show environment variables")
      print("* Include -inc to show included paths")
      print("* Include -plg to show PGLOG variables")

   print("********************************************************************\n")

   sys.exit(0)

#
# call main() to start program
#
if __name__ == "__main__": main()
